File: src/dotnet/AdventOfCode2021/AdventOfCode2021.Solutions/Day2.cs
namespace AdventOfCode2021.Solutions;

public partial class Day2
{
    private record struct Coordinates(int Depth, int Position);

    private record struct ExtendedCoordinates(int Depth, int Position, int Aim);

    public int CalculatePartOne()
    {
        Console.WriteLine("Day 2. Part 1");

        var values = Input.Split(Environment.NewLine);

        var lastCoordinates = values.Aggregate(new Coordinates(0, 0), (coords, command) =>
        {
            var parts = command.Split(" ");
            var action = parts[0];
            var amount = int.Parse(parts[1]);

            return action switch
            {
                "forward" => coords with { Position = coords.Position + amount },
                "up" => coords with { Depth = coords.Depth - amount },
                "down" => coords with { Depth = coords.Depth + amount },
                _ => throw new Exception($"Invalid command {action}")
            };
        });

        return lastCoordinates.Depth * lastCoordinates.Position;
    }

    public int CalculatePartTwo()
    {
        Console.WriteLine("Day 2. Part 2");

        var values = Input.Split(Environment.NewLine);

        var lastCoordinates = values.Aggregate(new ExtendedCoordinates(0, 0, 0), (coords, command) =>
        {
            var parts = command.Split(" ");
            var action = parts[0];
            var amount = int.Parse(parts[1]);

            return action switch
            {
                "forward" => coords with { Position = coords.Position + amount, Depth = coords.Depth + coords.Aim * amount },
                "up" => coords with { Aim = coords.Aim - amount },
                "down" => coords with { Aim = coords.Aim + amount },
                _ => throw new Exception($"Invalid command {action}")
            };
        });

        return lastCoordinates.Depth * lastCoordinates.Position;
    }
}

File: src/kotlin/app/src/main/kotlin/ru/renev/adventofcode2021/Day2.kt
package ru.renev.adventofcode2021

class Day2 {
    fun part1(): Int {
        println("Day 2 part 1")

        val parts = Day2Input.values.split("\n")
        var pos = 0
        var depth = 0
        for (command in parts) {
            val (action, amount) = command.split(" ")
            val amountValue = amount.toInt()

            when (action) {
                "forward" -> pos += amountValue
                "down" -> depth += amountValue
                "up" -> depth -= amountValue
                else -> throw RuntimeException("Invalid action $action")
            }
        }

        return pos * depth
    }

    fun part2(): Int {
        println("Day 2 part 2")

        val parts = Day2Input.values.split("\n")
        var pos = 0
        var depth = 0
        var aim = 0
        for (command in parts) {
            val (action, amount) = command.split(" ")
            val amountValue = amount.toInt()

            when (action) {
                "forward" -> {
                    pos += amountValue
                    depth += aim * amountValue
                }
                "down" -> aim += amountValue
                "up" -> aim -= amountValue
                else -> throw RuntimeException("Invalid action $action")
            }
        }

        return pos * depth
    }
}
